src/ripple/conditions/prefix_sha256.h

```diff
@@ -0,0 +1,204 @@
+
+#ifndef RIPPLE_CONDITIONS_PREFIX_SHA256_H
+#define RIPPLE_CONDITIONS_PREFIX_SHA256_H
+
+#include <ripple/basics/base_uint.h>
+#include <ripple/conditions/condition.h>
+#include <ripple/conditions/fulfillment.h>
+#include <ripple/conditions/impl/base64.h>
+#include <ripple/protocol/digest.h>
+#include <boost/regex.hpp>
+#include <boost/optional.hpp>
+#include <algorithm>
+#include <iterator>
+#include <memory>
+#include <vector>
+
+namespace ripple {
+namespace cryptoconditions {
+
+struct prefix_sha256_t final
+    : public fulfillment_t
+{
+    Buffer prefix_;
+    std::unique_ptr<fulfillment_t> subfulfillment_;
+
+public:
+    prefix_sha256_t () = default;
+
+    Buffer
+    payload() const override
+    {
+        auto subpayload = to_blob (*subfulfillment_);
+
+        // Prepend the prefix to the message:
+        Buffer b (subpayload.size() +
+            oer::predict_octetstring_size (prefix_.size()));
+
+        auto out = oer::encode_octetstring (
+            prefix_.size(),
+            prefix_.begin(),
+            prefix_.end(),
+            b.begin());
+
+        std::copy (
+            subpayload.begin(),
+            subpayload.end(),
+            out);
+
+        return b;
+    }
+
+    condition_t
+    condition() const override
+    {
+        auto const& sc = subcondition();
+        auto const blob = to_blob (sc);
+
+        auto const payloadSize = oer::predict_octetstring_size (
+            sc.fulfillment_length);
+
+        // FIXME: we can reduce copying here by incrementally
+        // hashing:
+        Buffer b (blob.size() +
+            oer::predict_octetstring_size (prefix_.size()));
+
+        auto out = oer::encode_octetstring (
+            prefix_.size(),
+            prefix_.begin(),
+            prefix_.end(),
+            b.begin());
+
+        std::copy (
+            blob.begin(),
+            blob.end(),
+            out);
+
+        sha256_hasher h;
+        h (b.data(), b.size());
+
+        condition_t cc;
+        cc.type = type();
+        cc.requires = requires();
+        cc.fulfillment_length = prefix_.size() + payloadSize;
+        cc.fingerprint = static_cast<sha256_hasher::result_type>(h);
+
+        return cc;
+    }
+
+    std::uint16_t
+    type () const override
+    {
+        return condition_prefix_sha256;
+    }
+
+    std::uint32_t
+    requires () const override
+    {
+        return
+            feature_sha256 |
+            feature_prefix |
+            subfulfillment_->requires();
+    }
+
+    bool
+    validate (Slice const& m) const override
+    {
+        if (!subfulfillment_)
+        {
+            std::cerr << "Failed because of no subf!\n";
+            return false;
+        }
+
+        // Prepend the prefix to the message:
+        Buffer b (prefix_.size() + m.size());
+
+        if (prefix_.size())
+            std::memcpy (b.data(), prefix_.data(), prefix_.size());
+
+        if (m.size())
+            std::memcpy (b.data(), m.data(), m.size());
```

Comment:
I know you caught this bug, but I'm leaving a note to make sure it doesn't get overlooked.